core/utils/syscallthread.h

```diff
@@ -0,0 +1,443 @@
+// and/or other materials provided with the distribution.
+//
+// * Neither the names of the copyright holders nor the names of their
+// contributors may be used to endorse or promote products derived from this
+// software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE
+// LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
+// CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
+// SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
+// INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
+// CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
+// ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+// POSSIBILITY OF SUCH DAMAGE.
+
+#ifndef BESS_UTILS_SYSCALLTHREAD_H
+#define BESS_UTILS_SYSCALLTHREAD_H
+
+#include <cassert>
+#include <cerrno>
+#include <thread>
+
+#include <nmmintrin.h>
+#include <signal.h>
+
+namespace bess {
+namespace utils {
+
+template <bool>
+class SyscallThread;  // forward
+
+bool CatchExitSignal();
+
+/*!
+ * At several points in bess we need to spin off threads that make
+ * blocking system calls, and be able to tell these threads to
+ * terminate.
+ *
+ * The actual code inside these threads varies a lot, but there
+ * is some common setup:
+ *
+ *  - establish a signal handler for the signal so that
+ *    the signal will interrupt blocking system calls
+ *  - establish reception of that signal in the thread
+ *  - send such a signal from outside the thread
+ *
+ * Signal handlers are themselves global across all threads, so
+ * the signal-handler-establishing can be done just once.
+ *
+ * We also pick which signal is to be used to interrupt the
+ * blocking system call, and provide a function to send that
+ * signal.
+ */
+
+class ExitSigMask {
+ public:
+  ExitSigMask() {
+    sigfillset(&allmask_);  // complete blockage of everything
+    sigfillset(&sigmask_);  // block all except SIG_THREAD_EXIT
+    sigdelset(&sigmask_, SIG_THREAD_EXIT);
+  }
+  const sigset_t &GetAllMask() const { return allmask_; }
+  const sigset_t &GetExitMask() const { return sigmask_; }
+
+ private:
+  static constexpr int SIG_THREAD_EXIT = SIGUSR2;
+  friend SyscallThread<true>;
+  friend SyscallThread<false>;
+  friend bool CatchExitSignal();
+  sigset_t allmask_;
+  sigset_t sigmask_;
+
+  // We only catch the exit signal once, process wide, so we want
+  // a flag to tell if we've done that.  Likewise, the per-thread
+  // signal mask is the same across all such threads, so we only
+  // need one instance.
+  static bool signals_initialized_;
+};
+
```

Comment:
We may move this class `ExitSigMask` into `syscallthread.cc`,  if this class is meant to be only accessible and visible to SyscallThread. Then inside .cc, this class can become simpler form as we don't need access control as well.

Reply:
This is a good idea, if I can get it to work out.